RFC/src/RFC-0201_TariScript.md

```diff
@@ -191,93 +202,177 @@ pub struct TransactionOutput {
-commitments to the UTXO script hashes. This also means that the kernel signatures are calculated slightly differently:
-
-$$  
 \begin{align}
-  s_i &= r_i + e.\bigl(k_i + \mathrm{H}(C_i \Vert s_i) \bigr) \\\\
-  s_i.G &= r_i.G + e.\bigl(k_i + \mathrm{H}(C_i \Vert s_i) \bigr).G \\\\
-  s_i.G &= R_i + e.\bigl(P_i + \mathrm{H}(C_i \Vert s_i)\bigr) \\\\
+s_i &= r_i + e.\pm \bigl(k_i + k_{i'}\sigma_i \bigr) \\\\
+s_i.G &= r_i.G + e.\pm \bigl(k_i + k_{i'}\sigma_i \bigr).G \\\\
+      &= R_i + e.\pm \bigl(P_i + L_i\sigma_i \bigr) \\\\
+    s &= \sum s_i \\\\
+    s.G  &= \sum R_i + e.\bigl( \sum \pm P_i \pm \sum L_i\sigma_i \bigr) \\\\
+         &= \sum R_i + e.\bigl( X_s + \sum \pm L_i\sigma_i \bigr) \\\\
 \end{align}
 $$
 
-Summing the signatures, one can easily confirm that \\( X_s + \Lambda \\) signs the kernel correctly. The kernel offset
+The exercise above demonstrates that $$x_s + \sum \pm L_i\sigma_i$$ signs the kernel correctly. The kernel offset
 is not included in this treatment, but it does not affect the result. One of the input commitments will be offset by a
 value selected by the sender and provided with the transaction data as usual. The signatures will still validate as
 usual, and the kernel offset will correct the overall excess balance.
 
 The same treatment extends to be block validation check. Note that since the individual kernel excesses can still be
-summed to obtain the overall block balance, the de-association of kernels and their outputs is maintained.
+summed to obtain the overall block balance.
+Furthermore, at the block aggregation level, it is non-trivial to assign a script hash to a given UTXO, and so the
+dis-association of kernels and their outputs is maintained.
+
+### Additional consensus rules
+
+In addition to the changes given above, there are two additional consensus rule changes for transaction and block validation.
+
+For every valid block or transaction,
+
+1. Every script hash in the kernel MUST be a default script hash or match a script in the script section. If there are duplicate script hashes in the kernel, then only one script MUST be presented in the script section.
+2. The script, with it's input data, if any, MUST execute successfully, i.e. without any errors. After execution, the stack MUST have exactly one element and its value MUST be exactly zero.
 
 ### Checking the requirements
 
-Miners cannot modify the script hash, because it is committed to in the public excess value. Moreover, the \\( C_i \Vert
-s_i \\) pair is committed to, so miners can't, for example swap script hashes on commitments to keep the overall excess
-the same but still manipulate specific outputs.
+Let's check whether the scripts are suitably committed on the blockchain, whether there's any malleability from miners or transaction participants and whether a new syncing node can verify the correctness of the chain.
+
+### Malleability
```

Comment:
I think here is another malleability attack here. 

While Bob might not be able to replace ` k_{a'}\sigma_a.G`. He can swop it out. 
So he creates a new tx, with `hat{C} &= k_i.G + k_{b'}\sigma_b.G + v_b.H `
But v_b has very little value. So when he published the combined tx. He swaps `k_{b'}\sigma_b.G` with `k_{a'}\sigma_a.G`.

With the current  scheme no one should be able to detect that happend.

Reply:
Bob doesn't know the value of `k_{a'}\sigma_a.G` so how can he swap them out?

Even if it were possible, the range proof must be generated with `k_a + k_a'σ_a` so any modification to the commitment invalidates the range proof; and to recreate a valid proof you need to know both secrets.